### PR TITLE
Speedup

### PR DESCRIPTION
**Pull request by nirtiac**
Speedups for everything except TabularDataset, that still needs some fixing. Note, many of the real speedups will have to come from user management of their resources given their data, I'm about to post a wiki article on that.

### CODE REVIEW

**Comment by nirtiac** [comment]
Potential serious issue with CUDNN optimization https://stackoverflow.com/questions/48445942/pytorch-training-with-gpu-gives-worse-error-than-training-the-same-thing-with-c

File: vulcanai/models/basenetwork.py [hunk 1]
```diff
@@ -872,32 +873,36 @@ def forward_pass(self, data_loader, transform_outputs=False,
...
-        pred_collector = torch.tensor([], dtype=dtype, device=self.device)
+        pred_collector = None
         for data, _ in data_loader:
+            data = set_tensor_device(data, device=self.device)
             # Get raw network output
-            raw_outputs = self(data)
+            raw_outputs = self(data).cpu().detach().numpy()
             if self._num_classes:
                 if self._final_transform:
                     predictions = self._final_transform(raw_outputs)
```
Comment (rfratila) [inline vulcanai/models/basenetwork.py]: Wouldn't it be better to keep the torch tensor through the final transform?

File: vulcanai/models/metrics.py [hunk 2]
```diff
@@ -62,6 +62,8 @@ def get_score(targets, predictions, metrics='accuracy', average=None,
         functions_without_average_parameter = ["get_accuracy", "get_mse"]
 
         if isinstance(targets, torch.Tensor):
+            # TODO: this is not necessary if targets was created as numpy on
+            # the CPU and then converted using from_numpy
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: is there way to do a check for this?

File: vulcanai/models/metrics.py [hunk 3]
```diff
@@ -598,6 +600,7 @@ def _run_test_single_continuous(network, data_loader, transform_outputs,
             results : dict
 
         """
+        # TODO: this may not work with current TabularDataset
         targets = np.array([v[1] for v in data_loader.dataset])
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: If you're referring to the matrix stored version within Tabular Dataset, wouldn't that be easier to extract?

File: vulcanai/datasets/tabulardataset.py [hunk 4]
```diff
@@ -130,14 +133,17 @@ def __getitem__(self, idx):
...
-            xs = torch.tensor(xs, dtype=torch.float)
-            y = self.df[[self.label_column]].iloc[[idx]].values.tolist()[0][0]
+            if self.all_xs is None:
+                self.all_xs = self.df.drop(self.label_column,
+                              axis=1)
+            if self.all_y is None:
+                self.all_y = self.df[[self.label_column]]
+            xs = torch.from_numpy(self.all_xs.iloc[[idx]].values[0]).float()
+            y = self.all_y.iloc[[idx]].values.tolist()[0][0]
             y = torch.tensor(y, dtype=torch.long)
```
Comment (rfratila) [inline vulcanai/datasets/tabulardataset.py]: why not also do a `torch.from_numpy` here as well?

File: vulcanai/models/basenetwork.py (hunk 1 above)
Comment (nirtiac) [reply vulcanai/models/basenetwork.py]: because it could be a numpy function, no?

File: vulcanai/models/metrics.py (hunk 3 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: yeah, this'll be resolved as long as we convert it to that. I'll take the comment out.

File: vulcanai/datasets/tabulardataset.py (hunk 4 above)
Comment (nirtiac) [reply vulcanai/datasets/tabulardataset.py]: because it was giving me all sorts of weird values due to it being a single value (shape was wrong). doesn't matter anyways though because we're going to scrap all of this once I  push the new tabulardataset.

File: vulcanai/models/metrics.py (hunk 2 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: hm good point will look into

File: vulcanai/models/basenetwork.py (hunk 1 above)
Comment (rfratila) [reply vulcanai/models/basenetwork.py]: I guess it could be but this is where we’d apply a soft max for instance and maybe it’s better to keep as torch for optimization?

Comment (nirtiac) [reply vulcanai/models/basenetwork.py]: Only in the case of nn.CrossEntropyLoss?

File: vulcanai/models/metrics.py (hunk 2 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: As far as I can tell it appears not @rfratila, although this was interesting reading https://stackoverflow.com/questions/48482787/pytorch-memory-model-torch-from-numpy-vs-torch-tensor

File: vulcanai/models/basenetwork.py (hunk 1 above)
Comment (rfratila) [reply vulcanai/models/basenetwork.py]: perhaps not only for CrossEntropy but should we restrict users to only Pytorch functions or numpy functions? curious to hear your thoughts on this - I think it really depends on the scope of those final transforms and how complex but still

File: vulcanai/models/metrics.py (hunk 2 above)
Comment (rfratila) [reply vulcanai/models/metrics.py]: perhaps we should modify the condition to make it more useful then? such as checking if on a GPU device?

Comment (nirtiac) [reply vulcanai/models/metrics.py]: So I think we want to use .detach regardless cause it adds in requires_grad=False, and then as far as I can tell .cpu is a no-op if it's already on the the CPU, and then we'll have to convert to numpy in any case.

File: vulcanai/models/basenetwork.py (hunk 1 above)
Comment (nirtiac) [reply vulcanai/models/basenetwork.py]: uhhhhhhhmmmmmm yeah I don't really know. Personally, I think we should choose ease of use over speed, particularly as in this case I don't see a huge GPU advantage? it's a transform after training.

Comment (nirtiac) [reply vulcanai/models/basenetwork.py]: like, with it this way they can give either kind of function. I definitely want to be able to pass something like numpy.round() or whatever.

File: vulcanai/models/metrics.py (hunk 2 above)
Comment (rfratila) [reply vulcanai/models/metrics.py]: sounds good. the condition can then make it support both torch tensors and numpy matrices

File: vulcanai/models/basenetwork.py (hunk 1 above)
Comment (rfratila) [reply vulcanai/models/basenetwork.py]: You're right - it won't slow down training but then depending on the function passed it may slow it down during inference. To put to ease the speed issue, would you be able to experiment the numpy.round() example and compared with a comparative torch.round() and see what, if any, slow down there is (and potentially for softmax)?
